addons/website_portal_sale/models/sale_order.py

```diff
@@ -24,25 +41,27 @@ def _compute_payment_transaction_count(self):
             order.payment_transaction_count = mapped_data.get(order.id, 0)
 
     @api.multi
-    def get_access_action(self):
-        """ Instead of the classic form view, redirect to the online quote for
+    def get_access_action(self, access_token=None):
+        """ Instead of the classic form view, redirect to the online order for
         portal users that have access to a confirmed order. """
         # TDE note: read access on sales order to portal users granted to followed sales orders
         self.ensure_one()
         if self.state == 'cancel' or (self.state == 'draft' and not self.env.context.get('mark_so_as_sent')):
             return super(SaleOrder, self).get_access_action()
-        if self.env.user.share or self.env.context.get('force_website'):
+        if self.env.user.share or access_token or self.env.context.get('force_website'):
+            url_action = {
+                'type': 'ir.actions.act_url',
+                'url': '/my/orders/%s?access_token=%s' % (self.id, self.access_token),
+                'target': 'self',
+                'res_id': self.id,
+            }
             try:
                 self.check_access_rule('read')
             except exceptions.AccessError:
-                pass
+                if self.access_token == access_token:
```

Comment:
Don't do sensitive string comparison with `==`
Import consteq (`from odoo.tools import consteq`) then use `consteq(correct_token, provided_token)`, to prevent timing attacks.